File: leetcode/buildTree.cpp
/**********************************************
  > File Name		: buildTree.cpp
  > Author		    : lunar
  > Email			: lunar_ubuntu@qq.com
  > Created Time	: Fri 15 Apr 2022 04:33:25 PM CST
  > Location        : Shanghai
  > Copyright@ https://github.com/xiaoqixian
 **********************************************/

#include <vector>
#include <stack>
using namespace std;

struct TreeNode {
    int val;
    TreeNode* left, *right;
    TreeNode(int val): val(val), left(nullptr), right(nullptr) {}
};

class Solution {
public:
    TreeNode* buildTree(vector<int>& preorder, vector<int>& inorder) {
        if (preorder.empty()) return nullptr;

        return _buildTree(preorder, inorder, make_pair(0, preorder.size()-1), 
                make_pair(0, inorder.size()-1));
    }

private:
    TreeNode* _buildTree(vector<int>& preorder, vector<int>& inorder, 
            pair<int, int> preBound, pair<int, int> inBound) {
        TreeNode* root = new TreeNode(preorder[preBound.first]);

        if (preBound.first == preBound.second)
            return root;

        int inRootIndex = inBound.first, rootVal = preorder[preBound.first];
        while (inorder[inRootIndex] != rootVal)
            inRootIndex++;

        int leftTreeLength = 0;
        
        if (inRootIndex > inBound.first) {
            leftTreeLength = inRootIndex - inBound.first;
            root->left = _buildTree(preorder, inorder, 
                    make_pair(preBound.first + 1, preBound.first + leftTreeLength),
                    make_pair(inBound.first, inRootIndex - 1));
        }

        if (inRootIndex > inBound.second) {
            root->right = _buildTree(preorder, inorder, 
                    make_pair(preBound.first + leftTreeLength + 1, preBound.second),
                    (make_pair(inRootIndex + 1, inBound.second)));
        }

        return root;
    }
};

class Solution2 {
public:
    TreeNode* buildTree(vector<int>& preorder, vector<int>& inorder) {
        if (!preorder.size()) {
            return nullptr;
        }
        TreeNode* root = new TreeNode(preorder[0]);
        stack<TreeNode*> stk;
        stk.push(root);
        int inorderIndex = 0;
        for (int i = 1; i < preorder.size(); ++i) {
            int preorderVal = preorder[i];
            TreeNode* node = stk.top();
            if (node->val != inorder[inorderIndex]) {
                node->left = new TreeNode(preorderVal);
                stk.push(node->left);
            }
            else {
                while (!stk.empty() && stk.top()->val == inorder[inorderIndex]) {
                    node = stk.top();
                    stk.pop();
                    ++inorderIndex;
                }
                node->right = new TreeNode(preorderVal);
                stk.push(node->right);
            }
        }
        return root;
    }
};

File: leetcode/build_tree.rs
/**********************************************
  > File Name		: build_tree.rs
  > Author		    : lunar
  > Email			: lunar_ubuntu@qq.com
  > Created Time	: Fri 15 Apr 2022 02:02:52 PM CST
  > Location        : Shanghai
  > Copyright@ https://github.com/xiaoqixian
 **********************************************/

use std::rc::Rc;
use std::cell::RefCell;
// Definition for a binary tree node.
#[derive(Debug, PartialEq, Eq)]
pub struct TreeNode {
  pub val: i32,
  pub left: Option<Rc<RefCell<TreeNode>>>,
  pub right: Option<Rc<RefCell<TreeNode>>>,
}

impl TreeNode {
  #[inline]
  pub fn new(val: i32) -> Self {
    TreeNode {
      val,
      left: None,
      right: None
    }
  }
}

struct Solution;

/*
 * preorder: [root, [left child tree], [right child tree]]
 * inorder:  [[left child tree], root, [right child tree]] 
 */
impl Solution {
    pub fn build_tree(preorder: Vec<i32>, inorder: Vec<i32>) -> Option<Rc<RefCell<TreeNode>>> {
        if preorder.is_empty() {return None;}

        Self::_build_tree(&preorder, &inorder, (0, preorder.len()-1), (0, inorder.len()-1))
    }

    fn _build_tree(preorder: &Vec<i32>, inorder: &Vec<i32>, pre_bound: (usize, usize), 
        in_bound: (usize, usize)) -> Option<Rc<RefCell<TreeNode>>> {
        
        let root = Rc::new(RefCell::new(TreeNode::new(preorder[pre_bound.0])));

        if pre_bound.0 == pre_bound.1 {return Some(root);}

        let in_root_index = {
            let mut i = in_bound.0;
            let root_val = preorder[pre_bound.0];
            while inorder[i] != root_val {i += 1;}
            i
        };

        let mut left_tree_len: usize = 0;
        if in_root_index > in_bound.0 {
            left_tree_len = in_root_index - in_bound.0;
            (*root.borrow_mut()).left = Self::_build_tree(preorder, inorder, 
                (pre_bound.0 + 1, pre_bound.0 + left_tree_len),
                (in_bound.0, in_root_index - 1));
        }

        if in_root_index < in_bound.1 {
            let right_tree_len = in_bound.1 - in_root_index;
            (*root.borrow_mut()).right = Self::_build_tree(preorder, inorder, 
                (pre_bound.0 + left_tree_len + 1, pre_bound.1),
                (in_root_index + 1, in_bound.1));
        }

        Some(root)
    }
}
